From 26cceff1a83442b80cb144602b7d6d1ee678bccb Mon Sep 17 00:00:00 2001
From: Tomas Henzl <thenzl@redhat.com>
Date: Mon, 8 Sep 2014 14:31:44 -0400
Subject: [scsi] hpsa: fix memory leak in hpsa_hba_mode_enabled

Message-id: <1410186734-19986-54-git-send-email-thenzl@redhat.com>
Patchwork-id: 92059
O-Subject: [RHEL6.6 e-stor PATCH 53/83] hpsa: fix memory leak in hpsa_hba_mode_enabled
Bugzilla: 1069185
RH-Acked-by: Maurizio Lombardi <mlombard@redhat.com>
RH-Acked-by: David Milburn <dmilburn@redhat.com>
RH-Acked-by: Ewan Milne <emilne@redhat.com>

And while we're at it fix a magic number

Signed-off-by: Joe Handzik <joseph.t.handzik@hp.com>
Signed-off-by: Stephen M. Cameron <scameron@beardog.cce.hp.com>
Signed-off-by: Christoph Hellwig <hch@lst.de>
(cherry picked from commit 6e8e8088aa9d2f0108dda6943aea8930be6f8f22)
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/hpsa.c b/drivers/scsi/hpsa.c
index eab81eb..7e924d0 100644
--- a/drivers/scsi/hpsa.c
+++ b/drivers/scsi/hpsa.c
@@ -2983,6 +2983,7 @@ u8 *figure_lunaddrbytes(struct ctlr_info *h, int raid_ctlr_position, int i,
 static int hpsa_hba_mode_enabled(struct ctlr_info *h)
 {
  int rc;
+ int hba_mode_enabled;
  struct bmic_controller_parameters *ctlr_params;
  ctlr_params = kzalloc(sizeof(struct bmic_controller_parameters),
   GFP_KERNEL);
@@ -2995,7 +2996,11 @@ static int hpsa_hba_mode_enabled(struct ctlr_info *h)
   kfree(ctlr_params);
   return 0;
  }
- return ctlr_params->nvram_flags & (1 << 3) ? 1 : 0;
+
+ hba_mode_enabled =
+  ((ctlr_params->nvram_flags & HBA_MODE_ENABLED_FLAG) != 0);
+ kfree(ctlr_params);
+ return hba_mode_enabled;
 }
 
 static void hpsa_update_scsi_devices(struct ctlr_info *h, int hostno)
diff --git a/drivers/scsi/hpsa.h b/drivers/scsi/hpsa.h
index 44235a2..1e3cf33 100644
--- a/drivers/scsi/hpsa.h
+++ b/drivers/scsi/hpsa.h
@@ -90,6 +90,7 @@ struct bmic_controller_parameters {
  u8   automatic_drive_slamming;
  u8   reserved1;
  u8   nvram_flags;
+#define HBA_MODE_ENABLED_FLAG (1 << 3)
  u8   cache_nvram_flags;
  u8   drive_config_flags;
  u16  reserved2;
-- 
1.7.1